packages/service-library/src/servicelib/async_utils.py

```diff
@@ -0,0 +1,113 @@
+import asyncio
+from functools import wraps
+from collections import deque
+import attr
+from typing import List, Dict
+
+
+@attr.s
+class Context:
+    in_queue: asyncio.Queue = attr.ib()
+    out_queue: asyncio.Queue = attr.ib()
+    initialized: bool = attr.ib()
+
+
+def run_serialized_in_context(target_args: List[str] = None):
+    """All request to function with same calling context will be serialized.
+
+    Example:
+
+    Given the following decorated function
+
+        @run_serialized_in_context(target_args=["param3", "param1"])
+        async def func(param1, param2, param3):
+            await asyncio.sleep(1)
+
+    The context will be formed by the values of the arguments "param3" and "param1".
+    The values must be serializable as they will be converted to string
+    and put together as storage key for the context.
+
+    The below calls will all run in a sequence:
+
+        functions = [
+            func(1, "something", 3),
+            func(1, "else", 3),
+            func(1, "here", 3),
+        ]
+        await asyncio.gather(*functions)
+
+    The following calls will run in parallel, because they have different contexts:
+
+        functions = [
+            func(1, "something", 3),
+            func(2, "else", 3),
+            func(3, "here", 3),
+        ]
+        await asyncio.gather(*functions)
+
+    """
+    target_args = [] if target_args is None else target_args
+
+    def internal(decorated_function):
+        contexts = {}
+
+        def get_context(args, kwargs: Dict) -> Context:
+            arg_names = decorated_function.__code__.co_varnames[
+                : decorated_function.__code__.co_argcount
+            ]
+            search_args = dict(zip(arg_names, args))
+            search_args.update(kwargs)
+
+            key_parts = deque()
+            for arg in target_args:
+                if arg not in search_args:
+                    message = (
+                        f"Expected '{arg}' in '{decorated_function.__name__}'"
+                        f" arguments. Got '{search_args}'"
+                    )
+                    raise ValueError(message)
+                key_parts.append(search_args[arg])
+
+            key = ":".join(map(str, key_parts))
```

Comment:
Taken your example above, then the context is "not conmutative"?? i.e. these will generate different context
`` run_serialized_in_context(target_args=["param3", "param1"]) !=  run_serialized_in_context(target_args=["param1", "param3"]) `` I guess this is not what you want, right?

Reply:
It is intended to be used like this. I went the extra mile to guarantee that regardless of the order in which the `target_args` are listed, the generated keys always puts the same values of "param3" and "param1" in the same order as declared in the target_args. This enables the generation of the intended context storage key.